### PR TITLE
build-args newline split

### PR DESCRIPTION
**Pull request by ruslandoga**
Fixes https://github.com/docker/build-push-action/issues/726

[`redhat-actions/buildah-build` seem to be splitting by newlines](https://github.com/redhat-actions/buildah-build/blob/4b8d36793be00cc9ee95b449fee9cd44a7a8c6e9/src/utils.ts#L131-L142)

### CODE REVIEW

**Review by crazy-max — CHANGES_REQUESTED** [review]
Thank you for contributing! That's SGTM but it appears your commit message is missing a DCO sign-off, causing the DCO check to fail. We require all commit messages to have a Signed-off-by line with your name and e-mail, which looks something like:

```
Signed-off-by: YourFirsName YourLastName <yourname@example.org>
```

Sorry for the hassle and let me know if you need help or more detailed instructions!

File: src/context.ts [hunk 1]
```diff
@@ -69,7 +69,7 @@ export async function getInputs(defaultContext: string): Promise<Inputs> {
   return {
     addHosts: await getInputList('add-hosts'),
     allow: await getInputList('allow'),
-    buildArgs: await getInputList('build-args', true),
+    buildArgs: getNaiveInputList('build-args'),
```
Comment (crazy-max) [inline src/context.ts:72]: I think we could just use the [`core.getMultilineInput` func](https://github.com/actions/toolkit/blob/6c1f9eaae833355a0b212b66c5f2e3ac366de185/packages/core/src/core.ts#L148-L161).

**Comment by ruslandoga** [comment]
`Signed-off-by: Ruslan Doga <rusl@n-do.ga>` didn't work, and the DCO action suggested using my username and github email.

I hope it's OK.

**Comment by crazy-max** [comment]
You forgot to build javascript artifacts with `docker buildx bake pre-checkin`: https://github.com/docker/build-push-action/blob/master/.github/CONTRIBUTING.md#submitting-a-pull-request

https://github.com/docker/build-push-action/actions/runs/3908784489/jobs/6680628395#step:3:308

```
#29 0.323 ERROR: Build result differs. Please build first with "docker buildx bake build"
#29 0.326  M dist/index.js
#29 0.326  M dist/index.js.map
```

**Comment by ruslandoga** [comment]
Done: https://github.com/docker/build-push-action/pull/736/commits/3bfdd83cd728e5d476cc7f96e54c98a5de85c9cf

**Review by crazy-max — APPROVED** [review]
LGTM thanks

**Comment by crazy-max** [comment]
@ruslandoga Hum I just recall it will break users that have multi line build arg value: https://github.com/docker/build-push-action/issues/374

**Comment by crazy-max** [comment]
@ruslandoga Does it work if you add quotes with current version?:

```yaml
- uses: docker/build-push-action@v3
  with:
    build-args: |
      "BUILD_INFO=${{ steps.meta.outputs.json }}"
```

**Comment by ruslandoga** [comment]
@crazy-max I think I've tried it and it didn't work because the quote have to be escaped in JSON. I'll try again.

**Comment by ruslandoga** [comment]
It doesn't work: https://github.com/ruslandoga/docker-build-info/actions/runs/3909945739/jobs/6681579929

**Comment by crazy-max** [comment]
And using [`toJSON` expression](https://docs.github.com/en/actions/learn-github-actions/expressions#tojson)?:

```yaml
- uses: docker/build-push-action@v3
  with:
    build-args: |
      BUILDINFO=${{ toJSON(steps.meta.outputs.json) }}
```

Alternatively I think we could just create env vars in metadata-action for [each output](https://github.com/docker/metadata-action#outputs) named:
* `DOCKER_METADATA_VERSION`
* ...
* `DOCKER_METADATA_JSON`

So you could just set:

```yaml
- uses: docker/build-push-action@v3
  with:
    build-args: |
      DOCKER_METADATA_JSON
```

WDYT?

**Comment by ruslandoga** [comment]
`toJSON` (https://github.com/ruslandoga/docker-build-info/pull/3) results in invalid (extra quoted) `BUILD_INFO` as well:

```console
/ # printenv BUILD_INFO
"{\"tags\":[\"ghcr.io/ruslandoga/docker-build-info:pr-3\",\"ghcr.io/ruslandoga/docker-build-info:toJson\",\"ghcr.io/ruslandoga/docker-build-info:ca87063\",\"ghcr.io/ruslandoga/docker-build-info:toJson-ca87063\"],\"labels\":{\"org.opencontainers.image.title\":\"docker-build-info\",\"org.opencontainers.image.description\":\"\",\"org.opencontainers.image.url\":\"https://github.com/ruslandoga/docker-build-info\",\"org.opencontainers.image.source\":\"https://github.com/ruslandoga/docker-build-info\",\"org.opencontainers.image.version\":\"pr-3\",\"org.opencontainers.image.created\":\"2023-01-13T09:45:48.327Z\",\"org.opencontainers.image.revision\":\"ca8706329ed9b7bfad723d9cc0cb96d08c9fba79\",\"org.opencontainers.image.licenses\":\"\"}}"
```

**Comment by crazy-max** [comment]
@ruslandoga I made some changes to the metadata action to stringify the JSON output and looks to work on my side: https://github.com/docker/metadata-action/pull/257/files#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3dR73

```yaml
- uses: docker/build-push-action@v3
  with:
    build-args:
      BUILD_METADATA=${{ steps.meta.outputs.json  }}
```

```dockerfile
FROM alpine
RUN apk add --no-cache jq
ARG BUILD_METADATA
RUN echo $BUILD_METADATA | jq
```

Can you try with `crazy-max/docker-metadata-action@env-output` and let me know?

**Comment by ruslandoga** [comment]
@crazy-max `echo` strips the outer quotes from its input. I think this issue should be tested with `printenv`

**Comment by crazy-max** [comment]
@ruslandoga Don't think it matters with `jq` but here it is: https://github.com/docker/metadata-action/actions/runs/3910565457/jobs/6682911373#step:4:207

**Comment by ruslandoga** [comment]
`jq` is irrelevant. The JSON should be valid for most non-cli parsers to work.

**Comment by ruslandoga** [comment]
Still doesn't work: https://github.com/ruslandoga/docker-build-info/pull/4/checks

**Comment by crazy-max** [comment]
Added a new step to check this and looks fine to me: https://github.com/docker/metadata-action/actions/runs/3910598664/jobs/6682982486#step:5:188

* workflow: https://github.com/crazy-max/docker-metadata-action/blob/04861f5102fc7042b8529e3f426856244a8cfba2/.github/workflows/ci.yml#L230-L237
* dockerfile: https://github.com/crazy-max/docker-metadata-action/blob/04861f5102fc7042b8529e3f426856244a8cfba2/test/json.Dockerfile

**Comment by ruslandoga** [comment]
Try adding more tags.

**Comment by crazy-max** [comment]
Ah yes indeed :disappointed: If you're willing to open a new PR with your suggestion to the CSV parse, feel free to do so. I will look later today at this issue if you have nothing yet.

**Pull request merged**